renderdoc/driver/d3d12/d3d12_counters.cpp

```diff
@@ -251,83 +264,157 @@ vector<CounterResult> D3D12Replay::FetchCounters(const vector<uint32_t> &counter
+  pipestatsQueryDesc.Count = maxEID;
+  pipestatsQueryDesc.NodeMask = 1;
+  pipestatsQueryDesc.Type = D3D12_QUERY_HEAP_TYPE_PIPELINE_STATISTICS;
+  ID3D12QueryHeap *pipestatsQueryHeap = NULL;
+  hr = m_pDevice->CreateQueryHeap(&pipestatsQueryDesc, __uuidof(pipestatsQueryHeap),
+                                  (void **)&pipestatsQueryHeap);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create pipeline statistics query heap %08x", hr);
+    return ret;
+  }
+
+  D3D12_QUERY_HEAP_DESC occlusionQueryDesc;
+  occlusionQueryDesc.Count = maxEID;
+  occlusionQueryDesc.NodeMask = 1;
+  occlusionQueryDesc.Type = D3D12_QUERY_HEAP_TYPE_OCCLUSION;
+  ID3D12QueryHeap *occlusionQueryHeap = NULL;
+  hr = m_pDevice->CreateQueryHeap(&occlusionQueryDesc, __uuidof(occlusionQueryHeap),
+                                  (void **)&occlusionQueryHeap);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create occlusion query heap %08x", hr);
+    return ret;
+  }
+
+  // TODO: Only supported with developer mode drivers!!!
   m_pDevice->SetStablePowerState(TRUE);
 
-  D3D12GPUTimerCallback cb(m_pDevice, this, queryHeap);
+  D3D12GPUTimerCallback cb(m_pDevice, this, timerQueryHeap, pipestatsQueryHeap, occlusionQueryHeap);
 
   // replay the events to perform all the queries
   m_pDevice->ReplayLog(0, maxEID, eReplay_Full);
 
+  // TODO:  Only supported with developer mode drivers!!!
   m_pDevice->SetStablePowerState(FALSE);
 
   ID3D12GraphicsCommandList *list = m_pDevice->GetNewList();
 
-  list->ResolveQueryData(queryHeap, D3D12_QUERY_TYPE_TIMESTAMP, 0, maxEID * 2, readbackBuf, 0);
+  list->ResolveQueryData(timerQueryHeap, D3D12_QUERY_TYPE_TIMESTAMP, 0, maxEID, readbackBuf, 0);
+  list->ResolveQueryData(pipestatsQueryHeap, D3D12_QUERY_TYPE_PIPELINE_STATISTICS, 0, maxEID,
+                         readbackBuf, sizeof(uint64_t) * 2 * maxEID);
+  list->ResolveQueryData(
+      occlusionQueryHeap, D3D12_QUERY_TYPE_OCCLUSION, 0, maxEID, readbackBuf,
+      sizeof(uint64_t) * 2 + sizeof(D3D12_QUERY_DATA_PIPELINE_STATISTICS) * maxEID);
 
   list->Close();
 
   m_pDevice->ExecuteLists();
   m_pDevice->FlushLists();
 
-  D3D12_RANGE range = {0, (SIZE_T)bufDesc.Width};
+  D3D12_RANGE range;
+  range.Begin = 0;
+  range.End = (SIZE_T)bufDesc.Width;
+
   void *data;
   hr = readbackBuf->Map(0, &range, &data);
   if(FAILED(hr))
   {
-    RDCERR("Failed to create query heap %08x", hr);
-    SAFE_RELEASE(queryHeap);
+    RDCERR("Failed to read timer query heap data %08x", hr);
     SAFE_RELEASE(readbackBuf);
+    SAFE_RELEASE(timerQueryHeap);
+    SAFE_RELEASE(pipestatsQueryHeap);
+    SAFE_RELEASE(occlusionQueryHeap);
     return ret;
   }
 
   uint64_t *timestamps = (uint64_t *)data;
+  D3D12_QUERY_DATA_PIPELINE_STATISTICS *pipelinestats =
+      (D3D12_QUERY_DATA_PIPELINE_STATISTICS *)&((uint64_t *)data)[maxEID * 2];
+  uint64_t *occlusion = (uint64_t *)&((
+      uint64_t *)data)[2 + (sizeof(D3D12_QUERY_DATA_PIPELINE_STATISTICS) / sizeof(uint64_t)) * maxEID];
```

Comment:
I think this would be more clearly expressed by changing `data` to `byte *data`, and then incrementing it before initialising each of these result pointers. Something like:

```cpp
uint64_t *timestamps = (uint64_t *)data;
data += maxEID * 2 * sizeof(uint64_t);
D3D12_QUERY_DATA_PIPELINE_STATISTICS *pipelinestats = (D3D12_QUERY_DATA_PIPELINE_STATISTICS *)data;
data += maxEID * sizeof(D3D12_QUERY_DATA_PIPELINE_STATISTICS);
uint64_t *occlusion = (uint64_t *)data;
```